### PR TITLE
Update symbolic Bundle Adjustment

### PR DESCRIPTION
**Pull request by servantf**
Update the symbolic bundle adjustment to be on par with updates of the classical bundle adjustment options.

### CODE REVIEW

File: src/aliceVision/sfm/liealgebra.hpp [hunk 1]
```diff
@@ -198,6 +198,77 @@ namespace SE3 {
...
+
+constexpr int rowFromCoordinates(int row, int col)
+{
+  return col * 4 + row;
+}
+
+/**
+ * Compute the inverse of source matrix
+ */
+inline Eigen::Matrix<double, 16, 16> d_inverse(const Matrix & source)
```
Comment (simogasp) [inline src/aliceVision/sfm/liealgebra.hpp:225]:
```suggestion
inline Eigen::Matrix<double, 16, 16> d_inverse(const Matrix4d & source)
```
If the size of the matrix (and the type) is to be known it's better to specify it

File: src/aliceVision/multiview/relativePose/Rotation3PSolver.cpp [hunk 2]
```diff
@@ -54,6 +54,56 @@ void Rotation3PSolver::solve(const Mat& p1, const Mat& p2, std::vector<robustEst
...
+
+    Mat::Index n = p2d.cols();
+    Mat p2dn(3, n);
+
+    for (int i = 0; i < n; i++)
+    {
+        Vec3 pn;
+        pn(0) = p2d(0, i);
+        pn(1) = p2d(1, i);
+        pn(2) = 1.0;
```
Comment (simogasp) [inline src/aliceVision/multiview/relativePose/Rotation3PSolver.cpp:71]: i think you can do
```suggestion
        Vec3 pn(p2d(0, i),  p2d(1, i), 1.0);
```

File: src/aliceVision/multiview/relativePose/Rotation3PSolver.cpp [hunk 3]
```diff
@@ -54,6 +54,56 @@ void Rotation3PSolver::solve(const Mat& p1, const Mat& p2, std::vector<robustEst
...
+void Rotation3PSolver34::solve(const Mat& p2d, const Mat& p3d, std::vector<robustEstimation::Mat34Model>& Rs) const
+{
+    assert(2 == p2d.rows());
+    assert(4 == p3d.rows());
+    assert(p2d.cols() == p3d.cols());
+
+    Mat::Index n = p2d.cols();
+    Mat p2dn(3, n);
+
+    for (int i = 0; i < n; i++)
```
Comment (simogasp) [inline src/aliceVision/multiview/relativePose/Rotation3PSolver.cpp:66]:
```suggestion
    for (Mat::Index i = 0; i < n; ++i)
```

File: src/aliceVision/multiview/relativePose/Rotation3PSolver.cpp [hunk 4]
```diff
@@ -54,6 +54,56 @@ void Rotation3PSolver::solve(const Mat& p1, const Mat& p2, std::vector<robustEst
...
+    }
+
+    Eigen::Matrix3d M = Eigen::Matrix3d::Zero();
+
+    for(int i = 0; i < n; ++i)
+    {
+        for(int k = 0; k < 3; ++k)
+        {
+            for(int l = 0; l < 3; ++l)
+            {
```
Comment (simogasp) [inline src/aliceVision/multiview/relativePose/Rotation3PSolver.cpp:85]:
```suggestion
    for(Mat::Index i = 0; i < n; ++i)
    {
        for(Mat::Index k = 0; k < 3; ++k)
        {
            for(Mat::Index l = 0; l < 3; ++l)
            {
```

File: src/aliceVision/multiview/relativePose/Rotation3PSolver.cpp [hunk 5]
```diff
@@ -54,6 +54,56 @@ void Rotation3PSolver::solve(const Mat& p1, const Mat& p2, std::vector<robustEst
...
+                M(k, l) += p2dn(k, i) * p3d(l, i);
+            }
+        }
+    }
+
+    Eigen::JacobiSVD<Eigen::Matrix3d> svd(M, Eigen::ComputeFullU | Eigen::ComputeFullV);
+    Eigen::Matrix3d U = svd.matrixU();
+    Eigen::Matrix3d V = svd.matrixV();
```
Comment (simogasp) [inline src/aliceVision/multiview/relativePose/Rotation3PSolver.cpp:93]:
```suggestion
    const Eigen::Matrix3d U = svd.matrixU();
    const Eigen::Matrix3d V = svd.matrixV();
```

File: src/aliceVision/multiview/relativePose/Rotation3PSolver.cpp [hunk 6]
```diff
@@ -54,6 +54,56 @@ void Rotation3PSolver::solve(const Mat& p1, const Mat& p2, std::vector<robustEst
...
+
+    Eigen::Matrix3d G = U * V.transpose();
+    Eigen::Matrix3d D = Eigen::Matrix3d::Identity();
+    D(2, 2) = 1.0 / G.determinant();
+
+    Mat34 R = Mat34::Zero();
+    R.block<3, 3>(0, 0) = U * D * V.transpose();
+
+    robustEstimation::Mat34Model ret(R);
```
Comment (simogasp) [inline src/aliceVision/multiview/relativePose/Rotation3PSolver.cpp:102]:
```suggestion
    const robustEstimation::Mat34Model ret(R);
```

File: src/aliceVision/multiview/relativePose/Rotation3PSolver.cpp [hunk 7]
```diff
@@ -54,6 +54,56 @@ void Rotation3PSolver::solve(const Mat& p1, const Mat& p2, std::vector<robustEst
...
+                M(k, l) += p2dn(k, i) * p3d(l, i);
+            }
+        }
+    }
+
+    Eigen::JacobiSVD<Eigen::Matrix3d> svd(M, Eigen::ComputeFullU | Eigen::ComputeFullV);
+    Eigen::Matrix3d U = svd.matrixU();
+    Eigen::Matrix3d V = svd.matrixV();
+
+    Eigen::Matrix3d G = U * V.transpose();
```
Comment (simogasp) [inline src/aliceVision/multiview/relativePose/Rotation3PSolver.cpp:95]:
```suggestion
    const Eigen::Matrix3d G = U * V.transpose();
```

File: src/aliceVision/sfm/BundleAdjustmentSymbolicCeres.cpp [hunk 8]
```diff
@@ -26,6 +26,158 @@ namespace sfm {
...
+    }
+    
+    size_t posDelta = 0;
+    if (!_lockFocal)
+    {
+      if (_lockFocalRatio)
+      {
+        x_plus_delta[0] = x[0] + delta[posDelta]; 
+        x_plus_delta[1] = x[1] + _focalRatio * delta[posDelta];
+        posDelta++;
```
Comment (simogasp) [inline src/aliceVision/sfm/BundleAdjustmentSymbolicCeres.cpp:82]:
```suggestion
        ++posDelta;
```
prefer preincrement

File: src/aliceVision/sfm/BundleAdjustmentSymbolicCeres.cpp [hunk 9]
```diff
@@ -64,7 +216,9 @@ class CostProjection : public ceres::CostFunction {
     SE3::Matrix T = cTr * rTo;
     geometry::Pose3 T_pose3(T.block<3,4>(0, 0));
 
-    Vec2 pt_est = _intrinsics->project(T_pose3, pt, true);
+    Vec4 pth = pt.homogeneous();
```
Comment (simogasp) [inline src/aliceVision/sfm/BundleAdjustmentSymbolicCeres.cpp:219]:
```suggestion
    const Vec4 pth = pt.homogeneous();
```

File: src/aliceVision/sfm/BundleAdjustmentSymbolicCeres.cpp [hunk 10]
```diff
@@ -109,6 +264,262 @@ class CostProjection : public ceres::CostFunction {
...
+    
+    /*Update intrinsics object with estimated parameters*/
+    size_t params_size = _intrinsics->getParams().size();
+    std::vector<double> params;
+    for (size_t param_id = 0; param_id < params_size; param_id++) {
+      params.push_back(parameter_intrinsics[param_id]);
+    }
+    _intrinsics->updateFromParams(params);
+
+    SE3::Matrix oTr = SE3::inverse(rTo);
```
Comment (simogasp) [inline src/aliceVision/sfm/BundleAdjustmentSymbolicCeres.cpp:301]:
```suggestion
    const SE3::Matrix oTr = SE3::inverse(rTo);
```

File: src/aliceVision/sfm/BundleAdjustmentSymbolicCeres.cpp [hunk 11]
```diff
@@ -109,6 +264,262 @@ class CostProjection : public ceres::CostFunction {
...
+    SE3::Matrix T = cTo * oTr;
+    geometry::Pose3 T_pose3(T.block<3,4>(0, 0));
+
+    Vec4 cartesianpt = _intrinsics->getCartesianfromSphericalCoordinates(pt);
+
+    Vec2 pt_est = _intrinsics->project(T_pose3, cartesianpt, true);
+    double scale = _measured.scale;
+    if (scale < 1e-12) scale = 1.0;
```
Comment (simogasp) [inline src/aliceVision/sfm/BundleAdjustmentSymbolicCeres.cpp:309]:
```suggestion
    const double scale = _measured.scale < 1e-12 ? 1.0 : _measured.scale;
```

File: src/aliceVision/sfm/BundleAdjustmentSymbolicCeres.cpp [hunk 12]
```diff
@@ -109,6 +264,262 @@ class CostProjection : public ceres::CostFunction {
...
+    }
+    _intrinsics->updateFromParams(params);
+
+    SE3::Matrix oTr = SE3::inverse(rTo);
+    SE3::Matrix T = cTo * oTr;
+    geometry::Pose3 T_pose3(T.block<3,4>(0, 0));
+
+    Vec4 cartesianpt = _intrinsics->getCartesianfromSphericalCoordinates(pt);
+
+    Vec2 pt_est = _intrinsics->project(T_pose3, cartesianpt, true);
```
Comment (simogasp) [inline src/aliceVision/sfm/BundleAdjustmentSymbolicCeres.cpp:307]:
```suggestion
    const geometry::Pose3 T_pose3(T.block<3,4>(0, 0));

    const Vec4 cartesianpt = _intrinsics->getCartesianfromSphericalCoordinates(pt);

    const Vec2 pt_est = _intrinsics->project(T_pose3, cartesianpt, true);
```

**Review by simogasp — CHANGES_REQUESTED** [review]

File: src/aliceVision/sfmData/Landmark.hpp [hunk 13]
```diff
@@ -63,6 +63,7 @@ struct Landmark
   feature::EImageDescriberType descType = feature::EImageDescriberType::UNINITIALIZED;
   Observations observations;
   image::RGBColor rgb = image::WHITE;    //!> the color associated to the point
+  IndexT referenceView = UndefinedIndexT;
```
Comment (fabiencastan) [inline src/aliceVision/sfmData/Landmark.hpp:66]: I'm not sure what's the meaning of this variable.
Could you add documentation?
Also, I don't see any place where this value is set.

File: src/aliceVision/sfm/BundleAdjustmentSymbolicCeres.cpp [hunk 14]
```diff
@@ -112,6 +112,226 @@ class CostProjection : public ceres::CostFunction {
   bool _withRig;
 };
 
+class CostProjectionRelative : public ceres::CostFunction {
```
Comment (fabiencastan) [inline src/aliceVision/sfm/BundleAdjustmentSymbolicCeres.cpp:267]: Needs doxygen doc

File: src/aliceVision/sfm/BundleAdjustmentSymbolicCeres.cpp [hunk 15]
```diff
@@ -112,6 +112,226 @@ class CostProjection : public ceres::CostFunction {
...
+    return true;
+  }
+
+private:
+  const sfmData::Observation & _measured;
+  const std::shared_ptr<camera::IntrinsicBase> _intrinsics;
+  bool _withRig;
+};
+
+class CostProjectionRelativeSimplified : public ceres::CostFunction {
```
Comment (fabiencastan) [inline src/aliceVision/sfm/BundleAdjustmentSymbolicCeres.cpp:352]: Needs doxygen doc

File: src/aliceVision/sfm/BundleAdjustmentSymbolicCeres.cpp [hunk 16]
```diff
@@ -112,6 +112,226 @@ class CostProjection : public ceres::CostFunction {
...
+    return true;
+  }
+
+private:
+  const sfmData::Observation & _measured;
+  const std::shared_ptr<camera::IntrinsicBase> _intrinsics;
+  bool _withRig;
+};
+
+class CostMotionConstantSpeed : public ceres::CostFunction {
```
Comment (fabiencastan) [inline src/aliceVision/sfm/BundleAdjustmentSymbolicCeres.cpp:417]: Needs doxygen doc

**Pull request merged**